### PR TITLE
Move big objects (>20k) from stack to heap.

### PR DESCRIPTION
**Pull request by vrabaud**
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch

### CODE REVIEW

**Comment by alalek** [comment]
This patch should go into [3.4 branch](https://github.com/opencv/opencv/wiki/Branches) first.
We will merge changes from 3.4 into master regularly (weekly/bi-weekly).

Please:
- change "base" branch of this PR: master => 3.4 (use "Edit" button near PR title)
- rebase your commits from master onto 3.4 branch. For example:
    `git rebase -i --onto upstream/3.4 upstream/master`
    (check list of your commits, save and quit (Esc + "wq" + Enter)
    where `upstream` is configured by following [this GitHub guide](https://help.github.com/articles/configuring-a-remote-for-a-fork/) and fetched (`git fetch upstream`).
- push rebased commits into source branch of your fork (with `--force` option)

Note: no needs to re-open PR, apply changes "inplace".

File: modules/flann/include/opencv2/flann/general.h [hunk 1]
```diff
@@ -31,6 +31,8 @@
 #ifndef OPENCV_FLANN_GENERAL_H_
 #define OPENCV_FLANN_GENERAL_H_
 
+#if CV_VERSION_MAJOR == 4
```
Comment (alalek) [inline modules/flann/include/opencv2/flann/general.h:34]: `<= 4` for 3.4 branch

File: modules/imgproc/src/imgwarp.cpp [hunk 2]
```diff
@@ -2167,7 +2167,8 @@ class WarpAffineInvoker :
     virtual void operator() (const Range& range) const CV_OVERRIDE
     {
         const int BLOCK_SZ = 64;
-        short XY[BLOCK_SZ*BLOCK_SZ*2], A[BLOCK_SZ*BLOCK_SZ];
+        AutoBuffer<short> __XY(BLOCK_SZ * BLOCK_SZ * 2), __A(BLOCK_SZ * BLOCK_SZ);
```
Comment (alalek) [inline modules/imgproc/src/imgwarp.cpp:2170]:
```.diff
-AutoBuffer<short>
+AutoBuffer<short, 0>
```

to reduce stack consumption (if data goes to heap anyway)

(all places)

**Comment by vrabaud** [comment]
Moved to 3.4 branch. Sorry, there was some FLANN stuff in there, that is a different PR. Fixed AutoBuffer<short, 0>

**Review by alalek — APPROVED** [review]
LGTM :+1:

**Pull request merged**